upstream-seed/src/lib.rs

```diff
@@ -95,6 +94,36 @@ async fn track_projects(client: peer::Peer, projects: Vec<Urn>) {
     }
 }
 
+/// Install signal handlers for SIGINT or SIGTERM and return when one of these signals is received.
+///
+/// Also starts a background task that exits the process if any of the signals is received after a
+/// grace period of ten seconds after the first signal.
+async fn install_signal_handler() {
+    const GRACE_PERIOD: std::time::Duration = std::time::Duration::from_secs(10);
+    let (shutdown_tx, shutdown_rx) = futures::channel::oneshot::channel();
+    tokio::spawn(async move {
+        use tokio::signal::unix::{signal, SignalKind};
+        let mut sig_int = signal(SignalKind::interrupt()).unwrap();
+        let mut sig_term = signal(SignalKind::terminate()).unwrap();
+
+        futures::future::select(sig_term.recv().boxed(), sig_int.recv().boxed()).await;
+        let _ = shutdown_tx.send(());
+
+        tracing::info!(
+            "shutting down. send SIGINT or SIGTERM again in 5 seconds to force shutdown"
+        );
+        let grace_period_end = std::time::Instant::now() + GRACE_PERIOD;
+        loop {
+            futures::future::select(sig_term.recv().boxed(), sig_int.recv().boxed()).await;
+            if std::time::Instant::now() > grace_period_end {
+                std::process::exit(5);
```

Comment:
What does error code `5` mean?

Reply:
Nothing in particular. I just chose an arbitrary exit code that isn’t 1 so that you could in theory distinguish the exit conditions via the code.